### PR TITLE
scryptTS

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: zh/course-03/lesson-01/chapter-12/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# 第十二章: 循环
```
Comment (xhliu) [inline zh/course-03/lesson-01/chapter-12/README.md:1]: Merge this with the previous chapter.

Chapter: Declare a winner or draw
https://beta.reactjs.org/learn/tutorial-tic-tac-toe#declaring-a-winner

File: zh/course-03/lesson-01/chapter-13/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# 第十三章: 总结
+
+恭喜你，你已经对 `scryptTS` 有了初步的了解，并编写了一个 `TicTacToe` 合约。
```
Comment (xhliu) [inline zh/course-03/lesson-01/chapter-13/README.md:3]: scryptTS -> sCrypt everywhere.

Also remove SV, just say Bitcoin.

**Review by xhliu — CHANGES_REQUESTED** [review]
👍 good draft.

Let's improve.

File: zh/course-03/lesson-01/chapter-02/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# 第二章: scryptTS 概述
+
+## TypeScript 嵌入式领域特定语言
+
+`scryptTS` 是一种基于 [Typescript](https://www.typescriptlang.org/) 的嵌入式领域特定语言 ([sDSL](https://en.wikipedia.org/wiki/Domain-specific_language#External_and_Embedded_Domain_Specific_Languages))，用于在比特币 SV 上编写智能合约。 嵌入式意味着它是另一种语言中的一种语言。 `scryptTS` 严格来说是 TypeScript 的子集，因此所有 `scryptTS` 代码都是有效的 TypeScript，但反之则不然。
+
+我们选择 TypeScript 作为宿主语言，因为它提供了一种简单、熟悉的语言 (JavaScript)，但具有类型安全性，可以轻松开始编写安全的智能合约。 如果您已经熟悉 TypeScript/JavaScript，则无需学习新的编程语言或工具。 如果您是 TypeScript 的新手，请观看这个有用的介绍视频。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-02/README.md:7]: Fixed

File: zh/course-03/lesson-01/chapter-02/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+
+## 比特币智能合约如何运作？
+
+比特币上的智能合约基于 UTXO 模型，这与以太坊使用的账户模型有很大不同。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-02/README.md:12]: removed

File: zh/course-03/lesson-01/chapter-04/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# 第四章: `scrypt-cli` 命令行工具
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-04/README.md:1]: fixed

File: zh/course-03/lesson-01/chapter-04/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+
+`scrypt-cli` 是使 `scryptTS` 的开发更快更容易的 CLI 工具。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-04/README.md:4]: use `npx scrypt-cli project my-proj` does't not need to install it.

File: zh/course-03/lesson-01/chapter-05/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  x: bigint;
+
+  constructor(x: bigint) {
+    super(...arguments);
+    this.x = x;
+  }
+
+  @method()
+  public unlock(x: bigint) {
+    assert(this.add(this.x, 1n) === x, "x is not the expected unlocking script");
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-05/README.md:20]: fIXED

File: zh/course-03/lesson-01/chapter-05/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  @method()
+  add(x0: bigint, x1:bigint) : bigint {
+    return x0 + x1;
+  }
+}
+```
+
+用 `@prop()` 和 `@method()` 装饰的类成员将最终出现在区块链上，因此必须是 TypeScript 的严格子集。 既没有装饰的成员是常规的 TypeScript，不会在链上。 `scryptTS` 的显着优势在于链上和链下代码都使用相同的语言编写：TypeScript。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-05/README.md:30]: fIXED

File: zh/course-03/lesson-01/chapter-06/README.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  x: bigint;
+
+  @prop(true)
+  y: FixedArray<boolean, 2>;
+
+  @prop(false)
+  z: ByteString;
+
+  @prop()
+  static readonly N: bigint = 3n;
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-06/README.md:44]: ADDED

File: zh/course-03/lesson-01/chapter-06/README.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  static readonly x: bigint;  // invalid, must be initialized at declaration time
+
+  @prop()
+  x: bigint = 3n;  // invalid, cannot be initialized at declaration time
+
+  @prop()
+  x;  // invalid, must show claim type
+}
+```
+
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-06/README.md:80]: DeleteD

File: zh/course-03/lesson-01/chapter-08/README.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+```
+
+
+## `@methed` 方法
+
+使用`@method` 装饰器来标记任何打算在链上运行的方法。
+
+与属性一样，智能合约也可以有两种方法：
+
+1. 带有 `@method` 装饰器：这些方法只能调用同样由 `@method` 装饰器装饰的方法。此外，只能访问由 `@prop` 装饰的属性或 CTC。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-08/README.md:19]: DeleteD

File: zh/course-03/lesson-01/chapter-08/README.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+public unlock(x: bigint) {
+    // only succeeds if x is 1
+    assert(this.add(this.x, 1n) === x, "unequal");
+}
+```
+
+注意： 
+
+1. 公共 `@methed` 方法至少需要包含一个 `assert()` 函数调用。
+2. 公共 `@methed` 方法最后一个函数调用必须是一个 `assert()` 函数调用。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-08/README.md:42]: add

File: zh/course-03/lesson-01/chapter-08/README.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+@method()
+public unlock(x: bigint) {
+    // only succeeds if x is 1
+    assert(this.add(this.x, 1n) === x, "unequal");
+}
+```
+
+注意： 
+
+1. 公共 `@methed` 方法至少需要包含一个 `assert()` 函数调用。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-08/README.md:41]: Delete

File: zh/course-03/lesson-01/chapter-10/README.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        this.count = count
+    }
+
+    @method()
+    public increment() {
+        // Increment counter value
+        this.count++
+
+        // construct outputs that contains latest contract state
+        let outputs = this.buildStateOutput(this.ctx.utxo.value);
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-10/README.md:92]: fixed

File: zh/course-03/lesson-01/chapter-04/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-04/README.md:4]: Fixed

File: zh/course-03/lesson-01/chapter-11/README.md [hunk 15]
```diff
@@ -0,0 +1,38 @@
+# 第十一章: 数组
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-11/README.md:1]: Fixed

File: zh/course-03/lesson-01/chapter-13/README.md (hunk 2 above)
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-13/README.md]: FIX

File: zh/course-03/lesson-01/chapter-12/README.md (hunk 1 above)
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-12/README.md]: FIX

File: zh/course-03/lesson-01/chapter-10/script-01/puzzle.ts [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    @prop()
+    alice: PubKey;
+    @prop()
+    bob: PubKey;
+
+    @prop(true)
+    is_alice_turn: boolean;
+
+    @prop(true)
+    board: FixedArray<bigint, typeof TicTacToe.BOARDLEN>;
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-10/script-01/puzzle.ts]: FIX

File: zh/course-03/lesson-01/chapter-07/README.md [hunk 17]
```diff
@@ -0,0 +1,68 @@
+# 第七章: 编译时常量
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-07/README.md]: DELETE

File: zh/course-03/lesson-01/chapter-09/script-01/solution.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+    static readonly BOARDLEN = 9;
+
+    @method()
+    public move(n: bigint, sig: Sig, amount: bigint): void {
+        // check position `n`
+        assert(n >= 0n && n < 9n);
+        // check signature `sig`
+        let player: PubKey = this.is_alice_turn ? this.alice : this.bob;
+        assert(this.checkSig(sig, player), `checkSig failed, sig: ${sig}, pubkey: ${player}`);
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-09/script-01/solution.ts]: REMOVED

File: zh/course-03/lesson-02/chapter-03/README.md [hunk 19]
```diff
@@ -0,0 +1,50 @@
+# 第三章: 实例化合约与部署合约
+
+## 实例化合约
+
+我们已经通过加载合约制品文件得到了合约类 `Tictactoe`。 当用户点击开始按钮后，通过此合约类来实例化合约。
+
+
+实例化合约类 `Tictactoe` 需要两个玩家 `alice` 和 `bob` 的公钥。可以通过 `Signer` 的 `getDefaultPubKey()` 接口来获取公钥。由只有一个钱包，我们假设 `alice` 和 `bob` 都使用同一个公钥。
+
+下面代码实例化合约并调用 `markAsGenesis()` 将合约标识为创世合约。
```
Comment (xhliu) [inline zh/course-03/lesson-02/chapter-03/README.md:10]: 标识为初始合约

File: zh/course-03/lesson-02/chapter-03/README.md [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+- `amount`: 表示部署时合约锁定的余额
+- `options`: 是可选参数，支持自定义找零地址，以及使用的 `address` 对应的私钥来签名交易。
+
+## 实战演习
+
+调用合约实例的 `deploy()` 方法来部署合约。在此之前需要先连接一个 `Signer` 或者 `Provider`。
+
+
+参考这个 [commit](https://github.com/sCrypt-Inc/tic-tac-toe/commit/3fef24179933381c765cd7744f33ecdb302bdc36)
```
Comment (xhliu) [inline zh/course-03/lesson-02/chapter-03/README.md:50]: Delete

**Review by xhliu — CHANGES_REQUESTED** [review]
👍 Almost there.

File: zh/course-01/lesson-02/chapter-02/README.md [hunk 21]
```diff
@@ -13,7 +13,7 @@ git clone -b webapp https://github.com/sCrypt-Inc/tic-tac-toe
 ##  scryptlib
 dApp 需要在前端页面与合约进行交互。 要做到这一点，我们将使用 sCrypt 官方发布的 JavaScript 库 —— [scryptlib](https://github.com/sCrypt-Inc/scryptlib).
 
-scryptlib 用于集成以 sCrypt 语言编写的 Bitcoin SV 智能合约的 Javascript/TypeScript SDK。
+scryptlib 用于集成以 sCrypt 语言编写的 Bitcoin 智能合约的 Javascript/TypeScript SDK。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-01/lesson-02/chapter-02/README.md:16]: ignore

File: zh/course-03/lesson-01/chapter-03/README.md [hunk 22]
```diff
@@ -9,3 +9,41 @@
 
 开发工具我们建议使用微软的 [Visual Studio Code](https://code.visualstudio.com/)
 
+
+## `scrypt-cli` 命令行工具
+
+`scrypt-cli` 是使 `sCrypt` 的开发更快更容易的 CLI 工具。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-03/README.md:15]: You don't need to install it globally, you can refer to the latest version directly using `npx scrypt-cli args...`

File: zh/course-03/lesson-01/chapter-07/README.md [hunk 23]
```diff
@@ -1,68 +1,36 @@
...
-注意，定义 CTC 时不能显示声明类型， 下面的定义是不是一个 CTC:
-
-```ts
-const N: number = 3; // invalid
-```
-
-## 一个 带有 `static` `readonly` 修饰符成员变量
-
-定义 CTC 时不能使用 `@prop()` 装饰器标记。同样不能显示声明类型。
+下面例子是比特币网络中最常见的合约：支付到公钥哈希。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-07/README.md:5]: added

File: zh/course-03/lesson-01/chapter-08/README.md [hunk 24]
```diff
@@ -1,80 +1,103 @@
...
+        this.count = count
+    }
+
+    @method()
+    public increment() {
+        // Increment counter value
+        this.count++
+
+        // construct single output that contains latest contract state
+        let outputs = this.buildStateOutput(this.ctx.utxo.value);
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-08/README.md:64]: fixed

File: zh/course-03/lesson-01/chapter-09/README.md [hunk 25]
```diff
@@ -1,36 +1,84 @@
...
+let done = false;
+for (let i = 0n; i < 3n; i++) {
+    if (!done) {
+        x = x * 2n
+        if (x >= 8n) {
+            done = true
+        }
     }
+}
+```
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-09/README.md:31]: yes, we have a example code:

image

File: zh/course-03/lesson-01/chapter-09/README.md [hunk 26]
```diff
@@ -1,36 +1,84 @@
...
+0, 1, 2
+3, 4, 5
+6, 7, 8
+0, 3, 6
+1, 4, 7
+2, 5, 8
+0, 4, 8
+2, 4, 6
+```
+
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-09/README.md:50]: good catch

File: zh/course-03/lesson-02/chapter-01/README.md [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+通过 `scryptTS` ，你就能方便地编译，测试，部署，调用合约了。
+
+`scryptTS` 可以通过 `npm` 安装。
+
+```bash
+// use NPM
+npm install scrypt-ts
+
+// use Yarn
+yarn add scryptlib
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-01/README.md:38]: fixed everywhere

File: zh/course-03/lesson-02/chapter-01/README.md [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+```
+
+安装完成后， 需对 Create React App 项目 做简单的配置修改。
+
+你可以参考这个 [commit]() 进行配置修改。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-01/README.md:43]: added

File: zh/course-03/lesson-02/chapter-01/README.md [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+
+## 编译合约
+
+首先将我们上一节课编写的 `TicTacToe` 合约放到 `src/contracts` 目录下。运行以下命令编译合约:
+
+```
+npx scrypt-cli compile
+```
+
+编译成功会在合约所在目录看到输出的合约制品文件 `tictactoe.json` 。合约制品文件可用于初始化合约类。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-01/README.md:60]: fixed

File: zh/course-03/lesson-02/chapter-01/README.md [hunk 30]
```diff
@@ -0,0 +1,77 @@
...
+
+通常我们需要在前端使用合约制品文件初始化合约类。之后合约类才可用于实例化合约。
+
+
+```ts
+import { TicTacToe } from './contracts/tictactoe';
+import artifact from './contracts/tictactoe.json';
+import { MergedArtifact } from 'scrypt-ts';
+TicTacToe.init(artifact as unknown as MergedArtifact);
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-01/README.md:69]: fixed , change to 
```ts
TicTacToe.init(artifact as MergedArtifact);
```

File: zh/course-03/lesson-02/chapter-01/README.md [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+```
+
+
+## 实战演习
+
+1. 在 `index.tsx` 中导入合约制品文件 `tictactoe.json`，并初始化合约类 `TicTacToe`
+
+你可以参考这个 [commit]() 进行修改。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-01/README.md:77]: Deleted

File: zh/course-03/lesson-02/chapter-02/README.md [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+<img src="https://github.com/sCrypt-Inc/image-hosting/blob/master/learn-scrypt-courses/testcoin.gif?raw=true" width="600">
+
+## Signer 与 Provider
+
+- `Signer` 是一个以某种方式直接或间接访问私钥的类，私钥可以签署消息和交易，以授权用户执行某种操作。
+
+- `Provider` 是区块链上非基于账户的操作的抽象，比如广播已经签署的交易，通常不直接参与签署交易或数据。
+
+
+通常钱包需要实现 `Signer` 的抽象接口。`Signer` 链接一个 `Provider` 来广播交易。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-02/README.md:16]: fixed

File: zh/course-03/lesson-02/chapter-02/README.md [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+
+安装完 **sensilet** 钱包后， 点击右上角的设置按钮，切换到测试网。然后复制你的钱包地址，去我们的 [水龙头](https://scrypt.io/#faucet) 领取测试网 BSV。
+
+<img src="https://github.com/sCrypt-Inc/image-hosting/blob/master/learn-scrypt-courses/testcoin.gif?raw=true" width="600">
+
+## Signer 与 Provider
+
+- `Signer` 是一个以某种方式直接或间接访问私钥的类，私钥可以签署消息和交易，以授权用户执行某种操作。
+
+- `Provider` 是区块链上非基于账户的操作的抽象，比如广播已经签署的交易，通常不直接参与签署交易或数据。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-02/README.md:13]: fixed

File: zh/course-03/lesson-02/chapter-02/README.md [hunk 34]
```diff
@@ -0,0 +1,63 @@
...
+
+在开始部署合约之前，我们需要先连接钱包。这里以 [sensilet](https://sensilet.com) 为例，介绍如何连接钱包。
+
+安装完 **sensilet** 钱包后， 点击右上角的设置按钮，切换到测试网。然后复制你的钱包地址，去我们的 [水龙头](https://scrypt.io/#faucet) 领取测试网 BSV。
+
+<img src="https://github.com/sCrypt-Inc/image-hosting/blob/master/learn-scrypt-courses/testcoin.gif?raw=true" width="600">
+
+## Signer 与 Provider
+
+- `Signer` 是一个以某种方式直接或间接访问私钥的类，私钥可以签署消息和交易，以授权用户执行某种操作。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-02/README.md:11]: fixed

File: zh/course-03/lesson-02/chapter-02/README.md [hunk 35]
```diff
@@ -0,0 +1,63 @@
...
+
+
+通常钱包需要实现 `Signer` 的抽象接口。`Signer` 链接一个 `Provider` 来广播交易。
+
+```ts
+signer.connect(provider);
+```
+
+
+## 连接 **sensilet** 钱包
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-02/README.md:23]: added

File: zh/course-03/lesson-02/chapter-02/README.md [hunk 36]
```diff
@@ -0,0 +1,63 @@
...
+      Connect Sensilet
+    </button>
+}
+```
+
+## 实战演习
+
+调用钱包的 `getConnectedTarget()` 接口请求连接钱包。
+
+参考这个 [commit](https://github.com/sCrypt-Inc/tic-tac-toe/commit/d3503a097c1a6ac1b28183372cd90f31868098a6)
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-02/README.md:63]: just the commit related to this chapter. i think it is useful.

File: zh/course-03/lesson-02/chapter-03/README.md [hunk 37]
```diff
@@ -0,0 +1,50 @@
+# 第三章: 实例化合约与部署合约
+
+## 实例化合约
+
+我们已经通过加载合约制品文件得到了合约类 `Tictactoe`。 当用户点击开始按钮后，通过此合约类来实例化合约。
+
+
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-03/README.md:7]: save to #20

File: zh/course-03/lesson-02/chapter-03/README.md (hunk 19 above)
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-03/README.md:10]: Fixed

File: zh/course-03/lesson-02/chapter-02/README.md (hunk 36 above)
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-02/README.md:63]: delete

File: zh/course-03/lesson-02/chapter-01/README.md [hunk 38]
```diff
@@ -0,0 +1,77 @@
...
+
+井字棋 dApp 的前端界面将使用 [Create React App](https://create-react-app.dev/) 来创建。
+
+你可以使用 
+
+```ts
+npx create-react-app tic-tac-toe --template typescript
+```
+
+或者直接克隆我们已经创建好的[tic-tac-toe 仓库](https://github.com/sCrypt-Inc/tic-tac-toe)。该仓库的 `onlyweb` 分支包含一个未集成 `sCrypt`， 只有前端代码的井字棋游戏。我们假设你已经具备前端开发的基础知识，因此我们不会花时间来介绍这部分代码。
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-01/README.md:15]: master branch contains all codes Include the code exchanged with the contract。

`onlyweb` just frontend code.

File: zh/course-03/lesson-02/chapter-04/README.md [hunk 39]
```diff
@@ -0,0 +1,47 @@
+# 第四章: 调用合约
+
+接下来就是开始下棋了，每下一步棋，就是对合约的一次调用，并触发合约状态的改变。Web 应用程序与合约的交互主要发生在这个阶段。
+
+调用合约需要完成以下工作:
+
+1. 通过当前实例的 `.next()` 方法创建一个新的合约实例。使用最新的游戏数据 `gameData` 更新新实例的状态。
+
+2. 通过 `bindTxBuilder` 方法为 `TicTacToe` 合约的 `move()` 方法添加交易构建器在交易构建器中构建调用合约的交易。
```
Comment (xhliu) [inline zh/course-03/lesson-02/chapter-04/README.md:9]: 交易构建器在交易构建器中
->
交易构建器, 在其中

File: zh/course-03/lesson-02/chapter-04/README.md [hunk 40]
```diff
@@ -0,0 +1,47 @@
...
+const {tx, next} = await current.methods.move(
+    n,
+    signature
+);
+
+// 4. save latest contract instance
+props.setContract(next?.instance)
+```
+
+至此，我们就完成了TicTacToe合约与webapp的交互。玩家的每一次下棋动作都会在区块链上产生相应的交易。
```
Comment (xhliu) [inline zh/course-03/lesson-02/chapter-04/README.md:39]: Delete
区块

File: zh/course-03/lesson-02/chapter-04/README.md [hunk 41]
```diff
@@ -0,0 +1,47 @@
...
+
+## 实战演习
+
+1. 游戏未结束时，需要添加一个包含游戏最新状态的输出。
+2. 调用合约的公共方法，如果公共方法的某个参数是 `Sig` 类型，则需要通过回调函数来返回签名。
+
+参考这个 [commit](https://github.com/sCrypt-Inc/tic-tac-toe/commit/eeb0a04d22e9d5f09155ea3dce41728dae8cb6bf)
```
Comment (xhliu) [inline zh/course-03/lesson-02/chapter-04/README.md]: Delete

File: zh/course-03/lesson-01/chapter-09/README.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (xhliu) [reply zh/course-03/lesson-01/chapter-09/README.md]: Replace the whole section with a simple example.
```ts
for(let i = 0; i < 9; i++) {
  ...
}
```

File: zh/course-03/lesson-02/chapter-01/README.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (xhliu) [reply zh/course-03/lesson-02/chapter-01/README.md]: Possible to remove MergedArtifact?

File: zh/course-03/lesson-02/chapter-04/README.md [hunk 42]
```diff
@@ -0,0 +1,47 @@
...
+
+// 2. bind a tx builder for move
+TicTacToe.bindTxBuilder('move', async (options: BuildMethodCallTxOptions<SmartContract>, n: bigint, sig: Sig) => {
+    ...
+}
+
+// 3. call contract.methods.move(...) to broadcast transaction
+const {tx, next} = await current.methods.move(
+    n,
+    signature
```
Comment (xhliu) [inline zh/course-03/lesson-02/chapter-04/README.md:32]: Explain how to get signature. It is non-trivial and I don't know how to get it.

File: zh/course-03/lesson-02/chapter-05/README.md [hunk 43]
```diff
@@ -0,0 +1,15 @@
...
+
+恭喜你，你已经成功完成前端页面与智能合约的交互。你已经具备开发dApp 的基础技能了。你太棒了！
+通过这两个节课，我们给你展示智能合约开发以及和智能合约交互的核心内容。如果你有兴趣，可以在这个井字棋dApp的基础上继续完善。比如：
+
+1. 为它搭建一个后端服务，支持不同的玩家进行对战
+2. 现在是在前端页面管理合约的 **UTXO** ，把它放到后端管理
+3. 接入更多的钱包
+4. 增加宫格，比如 5X5
+
+本课程的代码 [Tic-Tac-Toe: Bitcoin onchain App Tutorial](https://github.com/sCrypt-Inc/tic-tac-toe)
```
Comment (xhliu) [inline zh/course-03/lesson-02/chapter-05/README.md:11]: 代码
->
代码在

最后加上句号

File: zh/course-03/lesson-02/chapter-05/README.md [hunk 44]
```diff
@@ -0,0 +1,15 @@
...
+
+期待你构建出让人惊喜的dApp
```
Comment (xhliu) [inline zh/course-03/lesson-02/chapter-05/README.md]: 最后加上句号

**Review by xhliu — CHANGES_REQUESTED** [review]
More changes.

File: zh/course-03/lesson-01/chapter-05/README.md [hunk 45]
```diff
@@ -0,0 +1,100 @@
...
+let arr2D: FixedArray<FixedArray<bigint,  3> , 2>  = [[11n, 12n, 13n], [21n, 22n, 23n]];
+let d: FixedArray<bigint,  3> = arr2D[1];
+let idx = 2n;
+// read
+d[1] = a[Number(idx)];
+d[2] = arr2D[Number(idx)][1];
+// write
+a[Number(idx)] = 2n;
+// assign to an array variable
+a = arr2D[1];
```
Comment (xhliu) [inline zh/course-03/lesson-01/chapter-05/README.md:34]: Delete.

File: zh/course-03/lesson-01/chapter-05/README.md [hunk 46]
```diff
@@ -0,0 +1,100 @@
...
+
+4. `Sig`：签名
+
+
+其中， `PubKey` 和 `Sig` 是 `ByteString` 类型的子类型。如果你想了解更多的基本类型，可以查看[语言参考文档](https://scrypt.io/scrypt-ts/getting-started/how-to-write-a-contract#data-types) 
+
+
+## 数组
+
+数组是相同基本类型的固定大小的值列表。数组元素使用逗号分割。当你声明一个数组时，你必须这样声明它：
```
Comment (xhliu) [inline zh/course-03/lesson-01/chapter-05/README.md:20]: 你必须使用FixedArray<T, N>。T是类型，N是数组大小。

File: zh/course-03/lesson-01/chapter-05/README.md (hunk 45 above)
Comment (xhliu) [inline zh/course-03/lesson-01/chapter-05/README.md:34]: Delete

File: zh/course-03/lesson-02/chapter-04/README.md [hunk 47]
```diff
@@ -0,0 +1,47 @@
...
+4. 完成调用后，需要把新的合约实例保存起来，以便继续调用合约。
+
+以上几个步骤的代码实现:
+
+```ts
+// 1. create nextInstance
+const current = props.contract as TicTacToe;
+const nextInstance = current.next();
+// update nextInstance state
+Object.assign(nextInstance, Utils.toContractState(latestGameData));
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-04/README.md:22]: convert latest game data to contract state with Utils.toContractState, update nextInstance state

File: zh/course-03/lesson-02/chapter-04/README.md (hunk 39 above)
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-04/README.md:9]: fixed

File: zh/course-03/lesson-02/chapter-04/README.md (hunk 41 above)
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-04/README.md]: Deleted

File: zh/course-03/lesson-02/chapter-05/README.md (hunk 43 above)
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-05/README.md:11]: Fixed

File: zh/course-03/lesson-01/chapter-05/README.md (hunk 46 above)
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-05/README.md:20]: fixed

File: zh/course-03/lesson-01/chapter-05/README.md (hunk 45 above)
Comment (zhfnjust) [reply zh/course-03/lesson-01/chapter-05/README.md:34]: Deleted

File: zh/course-03/lesson-02/chapter-04/README.md (hunk 42 above)
Comment (zhfnjust) [reply zh/course-03/lesson-02/chapter-04/README.md]: 如果公共方法的某个参数是 `Sig` 类型，则需要通过回调函数来返回签名。合约连接的 `Signer` 会使用默认私钥进行签名，并将签名结果通过回调函数的参数 `sigResponses` 返回。使用 `findSigFrom()` 查找与公钥关联的签名。

```ts
const { tx: callTx } = await p2pkh.methods.unlock(
    (sigResponses: SignatureResponse[]) => findSigFrom(sigResponses, $publickey),
    $publickey
);
```

通过 `MethodCallOptions` 中 `sigRequiredAddress` 可以指定 `Signer` 使用哪个私钥进行签名。

```ts
const { tx: callTx } = await p2pkh.methods.unlock(
    (sigResponses: SignatureResponse[]) => findSigFrom(sigResponses, $publickey),
    $publickey,
    {
        sigRequiredAddress: $publickey.toAddress()
    } as MethodCallOptions<P2PKH>
);
```

**Pull request merged**